apicast/src/oauth/apicast_oauth.lua

```diff
@@ -1,26 +1,432 @@
-local get_token = require 'oauth.apicast_oauth.get_token'
-local callback = require 'oauth.apicast_oauth.authorized_callback'
-local authorize = require 'oauth.apicast_oauth.authorize'
 local setmetatable = setmetatable
+local random = require 'resty.random'
+local ts = require 'threescale_utils'
+local cjson = require 'cjson'
+local backend_client = require ('backend_client')
+local http_authorization = require 'resty.http_authorization'
+local env = require 'resty.env'
 
+local inspect = require 'inspect'
 local _M = {
   _VERSION = '0.1'
 }
 
 local mt = { __index = _M }
 
-function _M.new(service)
+-- Required params for each grant type and response type.
+_M.params = {
+  grant_type = {
+    authorization_code = {'client_id','redirect_uri','code'},
+    password = {'client_id','client_secret','username','password'},
+    client_credentials = {'client_id','client_secret'},
+    implicit = {'client_id','redirect_uri'}
+  },
+  response_type = {
+    code = {'client_id','redirect_uri'},
+    token = {'client_id','redirect_uri'}
+  }
+}
+
+function _M.new()
   return setmetatable(
     {
-      authorize = authorize.call,
-      callback = callback.call,
-      get_token = get_token.call,
-      service = service
     }, mt)
 end
 
+function _M.extract_params()
+  local params = {}
+  local auth
+
+  if ngx.var.http_authorization then
+    auth = http_authorization.new(ngx.var.http_authorization)
+  end
+
+  local method = ngx.req.get_method()
+  if not method == 'POST' then
+    _M.respond_with_error(400, 'invalid_HTTP_method')
+    return
+  end
+
+  ngx.req.read_body()
+  local body_params = ngx.req.get_post_args()
+
+  params.client_id = auth.userid or body_params.client_id
+  params.client_secret = auth.password or body_params.client_secret
+
+  params.grant_type = body_params.grant_type
+  params.code = body_params.code
+  params.redirect_uri = body_params.redirect_uri or body_params.redirect_url
+
+  return params
+end
+
 function _M.transform_credentials(_, credentials)
   return credentials
 end
 
-return _M
+function _M.respond_and_exit(status, body, headers)
+  -- TODO: is there a better way to populate the response headers?..
+  if headers then
+    for name,value in pairs(headers) do
+      ngx.header[name] = value
+    end
+  end
+
+  ngx.status = status
+  ngx.print(body)
+  ngx.exit(ngx.HTTP_OK)
+end
+
+function _M.respond_with_error(status, message, description)
+
+  --TODO: as per the RFC (https://tools.ietf.org/html/rfc6749#section-5.2) return WWW-Authenticate response header if 401
+  local headers = {
+    ['Content-Type'] = 'application/json;charset=UTF-8'
+  }
+  local err_msg = { error = message, error_description = description }
+  local body = cjson.encode(err_msg)
+  ngx.log(ngx.INFO, "error :" .. inspect(body))
```

Comment:
ah this was just some logging I forgot to remove. I'll clean up, thanks. Probably should add some more useful logging though...